### PR TITLE
1925 store b64 attachments

### PR DESCRIPTION
**Pull request by RamilGaripov**
Ticket: https://github.com/metriport/metriport-internal/issues/1925

### Description
- Added base-64 attachment extraction from XMLs during FHIR conversion
- The files are saved on S3, and the DocumentReferences are uploaded to the FHIR Server

### Testing

- Local
  - [x] Works locally when converting XMLs using the FHIR e2e test script
- Staging
  - [ ] Test the whole conversion flow:
    - [ ] Create 2 identical pts on 2 different facilities 
    - [ ] Upload fake XML with b64 attachments
    - [ ] Pull thru EHRs
    - [ ] See if documents appear in the doc list for pt

### Release Plan
- [ ] Merge this

### CODE REVIEW

File: packages/core/src/domain/document/upload.ts [hunk 1]
```diff
@@ -18,3 +19,16 @@ export function createUploadMetadataFilePath(
   const uploadFilePath = createUploadFilePath(cxId, patientId, docName);
   return `${uploadFilePath}_metadata.xml`;
 }
+
+export function createAttachmentUploadPath({
+  fileName,
+  ownId,
```
Comment (thomasyopes) [inline packages/core/src/domain/document/upload.ts:25]: bit a strange name for a param -- `id` or `attachmentId` might make more sense...

File: packages/core/src/external/cda/process-attachments.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+      type: "Patient",
+    },
+    extension: extensions,
+  };
+
+  const docRefs: DocumentReference[] = [];
+  const uploadDetails: UploadParams[] = [];
+
+  b64Attachments.map(async att => {
+    const act = att.act;
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:69]: act... as in `actual` -- may need more descriptive name.

File: packages/core/src/external/cda/process-attachments.ts [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+  return {
+    fileB64Contents,
+    mimeType,
+  };
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function fillDocumentReference(initDocRef: DocumentReference, act: any) {
+  const updatedDocRef = cloneDeep(initDocRef);
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:145]: I don't agree with this approach personally -- creating a new docref inline is likely better than deepcloning a shared one each cycle. And is a safer long term pattern in my mind. See how raf did a new bundle for the consolidated stuff.

File: packages/core/src/external/cda/process-attachments.ts [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+  fhirApi?: FhirClient;
+}) {
+  const { log } = out(`processAttachments - cxId ${cxId}, patientId: ${patientId}`);
+  const filePath = createFilePathFromFileName(fileName);
+
+  const extensions = [buildDocIdFhirExtension(filePath), getSourceExtension(medicalDataSource)]
+    .flat()
+    .filter(Boolean) as Extension[];
+
+  const initialDocRef: DocumentReference = {
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:54]: If you decide to stick to the cloneDeep approach, I'd make a comment that this is "shared" and will be cloned. But again, the fact that that comment is necessary is a sign it's a bad pattern IMO.

File: packages/core/src/external/cda/process-attachments.ts [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    const fileUrl = s3Utils.buildFileUrl(s3BucketName, fileKey);
+    log(`fileUrl: ${fileUrl}, and fileKey: ${fileKey}`);
+    const attachment: Attachment = {
+      contentType: fileDetails.mimeType,
+      url: fileUrl,
+      size: sizeInBytes(fileDetails.fileB64Contents),
+      title: fileKey,
+    };
+
+    if (docRef.date) attachment.creation;
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:90]: is creation a method? this line seemingly does nothing... might need a comment

File: packages/core/src/external/cda/remove-b64.ts [hunk 6]
```diff
@@ -1,13 +1,17 @@
...
 
 const notesTemplateId = "2.16.840.1.113883.10.20.22.2.65";
 const resultsTemplateId = "2.16.840.1.113883.10.20.22.2.3.1";
 const b64Representation = "B64";
 
-export function removeBase64PdfEntries(payloadRaw: string): string {
+export function removeBase64PdfEntries(payloadRaw: string): {
+  documentContents: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  b64Attachments: any[]; // TODO: improve typing
```
Comment (thomasyopes) [inline packages/core/src/external/cda/remove-b64.ts:13]: yea, was going to say, loooots of `any` in the file above.

File: packages/core/src/external/fhir/bundle/convert-to-transaction-bundle.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
+import { Bundle, BundleEntry } from "@medplum/fhirtypes";
```
Comment (thomasyopes) [inline packages/core/src/external/fhir/bundle/convert-to-transaction-bundle.ts:1]: Leaving a comment that this code is unchanged and just moved is helpful

File: packages/infra/lib/api-stack/fhir-converter-connector.ts [hunk 8]
```diff
@@ -136,6 +140,8 @@ export function createLambda({
       AXIOS_TIMEOUT_SECONDS: axiosTimeout.toSeconds().toString(),
       ...(config.lambdasSentryDSN ? { SENTRY_DSN: config.lambdasSentryDSN } : {}),
       API_URL: `http://${apiServiceDnsAddress}`,
+      FHIR_SERVER_URL: fhirServerUrl,
+      MEDICAL_DOCUMENTS_BUCKET_NAME: medicalDocumentsBucket.bucketName,
```
Comment (thomasyopes) [inline packages/infra/lib/api-stack/fhir-converter-connector.ts:144]: this doesn't appear to be used in the lambda that I can tell...

File: packages/core/src/domain/document/upload.ts [hunk 9]
```diff
@@ -18,3 +19,16 @@ export function createUploadMetadataFilePath(
   const uploadFilePath = createUploadFilePath(cxId, patientId, docName);
   return `${uploadFilePath}_metadata.xml`;
 }
+
+export function createAttachmentUploadPath({
```
Comment (thomasyopes) [inline packages/core/src/domain/document/upload.ts:23]: this may be a larger issue, but this is called `createAttachmentUploadPath` which implies it's producing the "full" S3 key (i.e. /x/y/z/filename)

but it's returning ``${fileName}_${ownId}${extension}`;` which "looks" like just a modified file name with no path.

If we use fileName to include the full path, then I'd rename this to be `createAttachmentUploadFileName`.

File: packages/core/src/external/cda/process-attachments.ts [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+const s3Utils = new S3Utils(Config.getAWSRegion());
+const s3BucketName = Config.getMedicalDocumentsBucketName();
+
+export async function processAttachments({
+  b64Attachments,
+  cxId,
+  patientId,
+  fileName,
+  medicalDataSource,
+  fhirApi,
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:37]: I'm wondering if we should create the fhirApi instead of passig it in... is there a pattern for passing the fhirApi in for the core stuff or does it normally fetch it itself (maybe consolidated has a pattern to follow).

File: packages/core/src/external/cda/process-attachments.ts [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  b64Attachments: any[];
+  cxId: string;
+  patientId: string;
+  fileName: string;
+  medicalDataSource?: string | undefined;
+  fhirApi?: FhirClient;
+}) {
+  const { log } = out(`processAttachments - cxId ${cxId}, patientId: ${patientId}`);
+  const filePath = createFilePathFromFileName(fileName);
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:48]: likely we are saving all these under the path of the original file right? might be easy to simply send the original full path, and simply make filePath = `fullPath.split('/')[0]` or something.

File: packages/core/src/external/cda/process-attachments.ts [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+
+  b64Attachments.map(async att => {
+    const act = att.act;
+
+    const fileDetails = getDetails(act.text);
+    const docRef = fillDocumentReference(initialDocRef, act);
+    if (!docRef.id) throw new Error("Missing ID in DocRef");
+
+    const fileKey = createAttachmentUploadPath({
+      fileName: filePath,
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:76]: yea... so I'd rename this arg to `filePath` per my comment above...

File: packages/core/src/external/cda/process-attachments.ts [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+
+  const docRefs: DocumentReference[] = [];
+  const uploadDetails: UploadParams[] = [];
+
+  b64Attachments.map(async att => {
+    const act = att.act;
+
+    const fileDetails = getDetails(act.text);
+    const docRef = fillDocumentReference(initialDocRef, act);
+    if (!docRef.id) throw new Error("Missing ID in DocRef");
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:73]: this can't be undefined based on the code...

File: packages/lambdas/src/sqs-to-converter.ts [hunk 14]
```diff
@@ -170,15 +173,30 @@ export async function handler(event: SQSEvent) {
...
           continue;
         }
-        const payloadNoB64 = removeBase64PdfEntries(payloadRaw);
+        const { documentContents: payloadNoB64, b64Attachments } =
+          removeBase64PdfEntries(payloadRaw);
+
+        log(`Extracted ${b64Attachments.length} B64 attachments`);
+        if (b64Attachments.length) {
+          const fhirApi = makeFhirApi(cxId, fhirUrl);
+          processAttachments({
```
Comment (thomasyopes) [inline packages/lambdas/src/sqs-to-converter.ts:190]: need an await in lambdas I think... otherwise this may cutoff during execution...

File: packages/core/src/domain/filename.ts [hunk 15]
```diff
@@ -62,3 +62,14 @@ export function parseFilePath(filePath: string): ParsedFileName | undefined {
   }
   return undefined;
 }
+
+export function createFilePathFromFileName(fileName: string): string {
+  const name = fileName.includes("/") ? fileName.split("/").pop() : fileName;
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/core/src/domain/filename.ts:67]: oh didnt realize that! thx

File: packages/core/src/domain/document/upload.ts (hunk 1 above)
Comment (RamilGaripov) [reply packages/core/src/domain/document/upload.ts:25]: fixed

File: packages/core/src/external/cda/process-attachments.ts [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+}: {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  b64Attachments: any[];
+  cxId: string;
+  patientId: string;
+  fileName: string;
+  medicalDataSource?: string | undefined;
+  fhirApi?: FhirClient;
+}) {
+  const { log } = out(`processAttachments - cxId ${cxId}, patientId: ${patientId}`);
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/core/src/external/cda/process-attachments.ts:47]: woops! good catch

File: packages/core/src/external/cda/process-attachments.ts (hunk 2 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/process-attachments.ts:69]: no, act as in act lol that's the name of the fields in CDAs

File: packages/core/src/external/cda/process-attachments.ts (hunk 4 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/process-attachments.ts:54]: ya, i changed the pattern

File: packages/core/src/external/cda/process-attachments.ts (hunk 5 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/process-attachments.ts:90]: woopsies again! Added the assignment statement. `creation` is an attribute of a `DocumentReference.content.attachment` in FHIR

File: packages/core/src/external/fhir/bundle/convert-to-transaction-bundle.ts (hunk 7 above)
Comment (RamilGaripov) [reply packages/core/src/external/fhir/bundle/convert-to-transaction-bundle.ts:1]: I literally made the comment and forgot to press "finish review"... Sorry!

File: packages/infra/lib/api-stack/fhir-converter-connector.ts (hunk 8 above)
Comment (RamilGaripov) [reply packages/infra/lib/api-stack/fhir-converter-connector.ts:144]: This is used in the underlying functions that the lambda calls. 
It's used when the attachments are extracted: 
- FHIR Server URL to upload DocumentReferences, so they can be pulled with other documents
- Bucket Name to upload the attachments to S3

File: packages/core/src/domain/document/upload.ts (hunk 9 above)
Comment (RamilGaripov) [reply packages/core/src/domain/document/upload.ts:23]: good point!

File: packages/core/src/domain/filename.ts [hunk 17]
```diff
@@ -62,3 +62,14 @@ export function parseFilePath(filePath: string): ParsedFileName | undefined {
   }
   return undefined;
 }
+
+export function createFilePathFromFileName(fileName: string): string {
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/core/src/domain/filename.ts:66]: ya, i think with how filenames are already structured in prod, this function is redundant. Locally, this made sense, but ya, not anymore

File: packages/core/src/external/cda/process-attachments.ts (hunk 10 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/process-attachments.ts:37]: ya, looks like we can just initialize it in the core func instead of lambda layer

File: packages/core/src/external/cda/process-attachments.ts (hunk 13 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/process-attachments.ts:73]: ya, but we're using FHIR types from medplum, and resource IDs are always optional. I was thinking to use an assertion operator (`!`), but decided to use an error instead. Lmk what's a better approach here in your opinion

File: packages/lambdas/src/sqs-to-converter.ts (hunk 14 above)
Comment (RamilGaripov) [reply packages/lambdas/src/sqs-to-converter.ts:190]: Oh.. I was actually leaving that async on purpose

File: packages/core/src/external/cda/remove-b64.ts (hunk 6 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/remove-b64.ts:13]: not sure if this is something we want to address right away tbh

File: packages/infra/lib/api-stack/fhir-converter-connector.ts (hunk 8 above)
Comment (thomasyopes) [reply packages/infra/lib/api-stack/fhir-converter-connector.ts:144]: Hmmm -- I feel like for core code that is executed via lambda, it might make sense to make have the pattern be to accept the env variable as a param then default to the config version (and bail if not set).

I feel like the disconnect between the lambda code and env variables is ripe for bugs, but it's possible we've done this quite a lot. At least for `const region = Config.getAWSRegion();` but it's automatically set by AWS so is relatively safe to call within core.

File: packages/lambdas/src/sqs-to-converter.ts (hunk 14 above)
Comment (thomasyopes) [reply packages/lambdas/src/sqs-to-converter.ts:187]: yea, but if it's async the lambda execution context might exit upon a return without this finishing and it'll just end immediately.

File: packages/core/src/external/cda/process-attachments.ts (hunk 11 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/process-attachments.ts]: already fixed

File: packages/lambdas/src/sqs-to-converter.ts (hunk 14 above)
Comment (RamilGaripov) [reply packages/lambdas/src/sqs-to-converter.ts:187]: cool, thx for letting me know! added

File: packages/infra/lib/api-stack/fhir-converter-connector.ts (hunk 8 above)
Comment (RamilGaripov) [reply packages/infra/lib/api-stack/fhir-converter-connector.ts:144]: hmmm, i actually like this more. This will also allow me to make the local utils script that we use for FHIR converter tests to not insert stuff on s3/fhir server, unless we want to

File: packages/core/src/external/cda/process-attachments.ts [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+import { sizeInBytes } from "../../util/string";
+import { uuidv4 } from "../../util/uuid-v7";
+import { S3Utils, UploadParams } from "../aws/s3";
+import { cqExtension } from "../carequality/extension";
+import { cwExtension } from "../commonwell/extension";
+import { makeFhirApi } from "../fhir/api/api-factory";
+import { convertCollectionBundleToTransactionBundle } from "../fhir/bundle/convert-to-transaction-bundle";
+import { buildDocIdFhirExtension } from "../fhir/shared/extensions/doc-id-extension";
+
+const s3Utils = new S3Utils(Config.getAWSRegion());
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:33]: I prefer this pattern:
 
 ```
const region = Config.getAWSRegion();

function getS3UtilsInstance(): S3Utils {
  return new S3Utils(region);
}
```

over `const s3Utils = new S3Utils(Config.getAWSRegion());` as it's a bit too global for me but up to you.

File: packages/core/src/external/cda/process-attachments.ts [hunk 19]
```diff
@@ -0,0 +1,232 @@
...
+
+    const fileKey = createAttachmentUploadFileName({
+      filePath,
+      attachmentId: docRef.id,
+      mimeType: fileDetails.mimeType,
+    });
+
+    const fileUrl = s3Utils.buildFileUrl(s3BucketName, fileKey);
+
+    if (!fileDetails.fileB64Contents) return;
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:77]: should we move this up to line 66 to avoid the unnecessary code in between?

File: packages/core/src/external/cda/process-attachments.ts [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+    resourceType: "Bundle",
+    type: "collection",
+    entry: docRefBundleEntries,
+  };
+
+  const transactionBundle = convertCollectionBundleToTransactionBundle({
+    fhirBundle: collectionBundle,
+  });
+
+  if (transactionBundle.entry?.length && fhirUrl && s3BucketName.length) {
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:111]: I'd maybe check fhir and s3 separately... maybe we want one but not the other?

also, `s3BucketName.length` is a strange -- are we allowing an empty string to signify no bucket? We should simply allow undefined.

File: packages/core/src/external/cda/process-attachments.ts [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+    log(`Starting uploads. ${JSON.stringify(transactionBundle)}`);
+    const fhirApi = makeFhirApi(cxId, fhirUrl);
+    await executeWithNetworkRetries(async () => await fhirApi.executeBatch(transactionBundle), {
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:114]: await Promise.all()...

I think line 118 needs an await per my comment here https://github.com/metriport/metriport/pull/2995#discussion_r1834870770

File: packages/core/src/external/cda/process-attachments.ts [hunk 22]
```diff
@@ -0,0 +1,232 @@
...
+      log,
+    });
+
+    executeAsynchronously(uploadDetails, async uploadParams => {
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:118]: need an await if this in an a lambda

File: packages/core/src/external/cda/process-attachments.ts [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+  b64Attachments,
+  cxId,
+  patientId,
+  filePath,
+  medicalDataSource,
+  s3BucketName,
+  fhirUrl,
+}: {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  b64Attachments: any[];
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:45]: could this be `CdaOriginalText`?

File: packages/core/src/external/cda/process-attachments.ts [hunk 24]
```diff
@@ -0,0 +1,232 @@
...
+  return undefined;
+}
+
+function getDate(time: EffectiveTimeLowHigh | undefined): string | undefined {
+  if (time?.low?._value) return buildDayjs(normalizeDateFromFhir(time.low._value)).toISOString();
+  if (time?.high?._value) return buildDayjs(normalizeDateFromFhir(time.high._value)).toISOString();
+  return undefined;
+}
+
+function normalizeDateFromFhir(dateString: string) {
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:200]: should we turn the input string into dayJs and then just format it? Would be cleaner and more stable I think.

Comment (thomasyopes) [reply packages/core/src/external/cda/process-attachments.ts:200]: maybe dayJs can't handle timezones?

File: packages/core/src/external/cda/remove-b64.ts [hunk 25]
```diff
@@ -1,45 +1,53 @@
...
+
   if (json.ClinicalDocument?.component?.structuredBody?.component) {
     //eslint-disable-next-line @typescript-eslint/no-explicit-any
     toArray(json.ClinicalDocument.component.structuredBody.component).forEach((comp: any) => {
       if (
         toArray(comp.section?.templateId).some(
           //eslint-disable-next-line @typescript-eslint/no-explicit-any
           (template: any) =>
-            template?.["@_root"] === notesTemplateId || template?.["@_root"] === resultsTemplateId
+            template?.["_root"] === notesTemplateId || template?.["_root"] === resultsTemplateId
```
Comment (thomasyopes) [inline packages/core/src/external/cda/remove-b64.ts:34]: why the change here? were we not flagging stuff before?

File: packages/core/src/external/cda/remove-b64.ts (hunk 6 above)
Comment (thomasyopes) [reply packages/core/src/external/cda/remove-b64.ts:13]: there's no XML general type or something?

File: packages/core/src/domain/document/upload.ts [hunk 26]
```diff
@@ -18,3 +19,16 @@ export function createUploadMetadataFilePath(
   const uploadFilePath = createUploadFilePath(cxId, patientId, docName);
   return `${uploadFilePath}_metadata.xml`;
 }
+
+export function createAttachmentUploadFileName({
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/core/src/domain/document/upload.ts:23]: lol

[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/core/src/domain/document/upload.ts:23]: fixed

File: packages/core/src/external/cda/process-attachments.ts (hunk 18 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/process-attachments.ts:33]: I added this. But if you look up this exact line:
```const s3Utils = new S3Utils(Config.getAWSRegion());```

you'll find like 10 more instances of this throughout the code base. But ya, never too late to start writing better code! 🫡

File: packages/core/src/external/cda/process-attachments.ts (hunk 19 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/process-attachments.ts:77]: 💯 ! done

File: packages/core/src/external/cda/process-attachments.ts (hunk 20 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/process-attachments.ts:111]: true that

File: packages/core/src/external/cda/process-attachments.ts (hunk 21 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/process-attachments.ts:114]: added

File: packages/core/src/external/cda/process-attachments.ts (hunk 24 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/process-attachments.ts:200]: i tried giving dayjs the string that XML has it in, i.e. "20230913230000+0000", and it's invalid. I tried using `.format` as well, and no luck 😒

File: packages/core/src/external/cda/remove-b64.ts (hunk 25 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/remove-b64.ts:34]: So all the CDA types that we have defined in the codebase come without the `@`. So to reuse those types, I'm removing this character from being added by the XML parser, and therefore from all the fields that used it

File: packages/core/src/external/cda/process-attachments.ts (hunk 20 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/process-attachments.ts]: I made some changes just now.. Not too happy with em myself, but I think it kinda makes sense. If bucket name is undefined, we will fill it in with an empty string, and then not do anything for s3 upload anyway, so that shouldn't affect anything. Besides, the lambda still requires that the bucket name is provided. So maybe this is fine?

File: packages/core/src/external/cda/process-attachments.ts [hunk 27]
```diff
@@ -0,0 +1,259 @@
...
+    const docRef = buildDocumentReference(patientId, extensions, act);
+    if (!docRef.id) throw new Error("Missing ID in DocRef");
+
+    const fileKey = createAttachmentUploadFilePath({
+      filePath,
+      attachmentId: docRef.id,
+      mimeType: fileDetails.mimeType,
+    });
+
+    const fileUrl = s3Utils.buildFileUrl(s3BucketName ?? "", fileKey);
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:82]: I'd check if bucket name exists, and simply have no URL in the attachment if no bucket exists. If url is not optional for Attachment, then I'd make *that* an empty string.

Generating the URL with an emptry string bucket name feels weird to me:

So:
`const fileUrl = s3BucketName ? s3Utils.buildFileUrl(s3BucketName, fileKey) : '';`

File: packages/core/src/external/cda/process-attachments.ts [hunk 28]
```diff
@@ -0,0 +1,259 @@
...
+      title: fileKey,
+    };
+
+    if (docRef.date) attachment.creation = docRef.date;
+    docRef.content = [{ attachment }];
+
+    docRefs.push(docRef);
+
+    const uploadParams = {
+      bucket: s3BucketName ?? "",
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:97]: only add this if the s3 bucket is defined -- we know this doesn't make sense otherwise.

File: packages/core/src/external/cda/process-attachments.ts [hunk 29]
```diff
@@ -0,0 +1,259 @@
...
+  };
+
+  const transactionBundle = convertCollectionBundleToTransactionBundle({
+    fhirBundle: collectionBundle,
+  });
+
+  if (transactionBundle.entry?.length) {
+    await Promise.all([
+      handleFhirUpload(transactionBundle, cxId, log, fhirUrl),
+      handleS3Upload(uploadDetails, log, s3Utils, s3BucketName),
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:120]: needs to be `s3BucketName ? handleS3Upload(uploadDetails, log, s3Utils, s3BucketName) : undefined` or something -- again, so reason to do anything s3 related if the bucket is not defined.

File: packages/core/src/external/cda/process-attachments.ts [hunk 30]
```diff
@@ -0,0 +1,259 @@
...
+    ]);
+  }
+  log(`Done...`);
+}
+
+async function handleFhirUpload(
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:126]: I'd move this into the promise all as `fhirUrl ? handleFhirUpload(transactionBundle, cxId, log, fhirUrl) : undefined`

I like the idea of moving the rest of it to a new method, but having it do nothing on undefined just a bit unncessary.

File: packages/core/src/external/cda/process-attachments.ts [hunk 31]
```diff
@@ -0,0 +1,259 @@
...
+  fhirUrl?: string
+): Promise<void> {
+  if (fhirUrl) {
+    log(`Transaction bundle: ${JSON.stringify(transactionBundle)}`);
+    const fhirApi = makeFhirApi(cxId, fhirUrl);
+    await executeWithNetworkRetries(() => fhirApi.executeBatch(transactionBundle), { log });
+  }
+}
+
+async function handleS3Upload(
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:139]: same as above.

File: packages/core/src/external/cda/process-attachments.ts [hunk 32]
```diff
@@ -0,0 +1,259 @@
...
+async function handleFhirUpload(
+  transactionBundle: Bundle<Resource>,
+  cxId: string,
+  log: typeof console.log,
+  fhirUrl?: string
+): Promise<void> {
+  if (fhirUrl) {
+    log(`Transaction bundle: ${JSON.stringify(transactionBundle)}`);
+    const fhirApi = makeFhirApi(cxId, fhirUrl);
+    await executeWithNetworkRetries(() => fhirApi.executeBatch(transactionBundle), { log });
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:135]: fhir api is not async?

File: packages/core/src/external/cda/remove-b64.ts (hunk 25 above)
Comment (thomasyopes) [reply packages/core/src/external/cda/remove-b64.ts:34]: Is this the only place that needs the change... seems like a good idea but want to make sure we're not relying on `@` anywhere else

File: packages/core/src/external/cda/process-attachments.ts (hunk 30 above)
Comment (thomasyopes) [reply packages/core/src/external/cda/process-attachments.ts:126]: Same with s3

File: packages/utils/src/fhir-converter/shared.ts [hunk 33]
```diff
@@ -84,6 +84,12 @@ export function getPatientIdFromFileName(fileName: string) {
   return parts.patientId;
 }
 
+export function getCxIdFromFileName(fileName: string) {
```
Comment (thomasyopes) [inline packages/utils/src/fhir-converter/shared.ts:87]: getCxIdFromFileNameOrCreate might be better -- the typing will make it look like you're going to throw if you can't find it (i.e. `getCxIdFromFileNameOrFail`)

File: packages/core/src/external/cda/process-attachments.ts [hunk 34]
```diff
@@ -0,0 +1,259 @@
...
+    entry: docRefBundleEntries,
+  };
+
+  const transactionBundle = convertCollectionBundleToTransactionBundle({
+    fhirBundle: collectionBundle,
+  });
+
+  if (transactionBundle.entry?.length) {
+    await Promise.all([
+      handleFhirUpload(transactionBundle, cxId, log, fhirUrl),
```
Comment (thomasyopes) [inline packages/core/src/external/cda/process-attachments.ts:119]: `hirUrl ? handleFhirUpload(transactionBundle, cxId, log, fhirUrl) : undefined` and `s3BucketName ? handleS3Upload(uploadDetails, log, s3Utils, s3BucketName) : undefined`

File: packages/core/src/external/cda/process-attachments.ts (hunk 30 above)
Comment (thomasyopes) [reply packages/core/src/external/cda/process-attachments.ts:126]: see above

File: packages/core/src/external/cda/remove-b64.ts (hunk 25 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/remove-b64.ts:34]: I'll do the due diligence to make sure it doesn't cause troubles downstream

File: packages/core/src/external/cda/process-attachments.ts (hunk 30 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/process-attachments.ts:126]: yup, good point!

File: packages/core/src/external/cda/process-attachments.ts (hunk 32 above)
Comment (RamilGaripov) [reply packages/core/src/external/cda/process-attachments.ts:135]: yes it does. Added

File: packages/core/src/external/cda/process-attachments.ts (hunk 24 above)
Comment (thomasyopes) [reply packages/core/src/external/cda/process-attachments.ts]: word

**Review by thomasyopes — APPROVED** [review]
